clang/lib/CIR/Dialect/Transforms/TargetLowering/ItaniumCXXABI.cpp

```diff
@@ -478,6 +482,61 @@ mlir::Value ItaniumCXXABI::lowerDataMemberCmp(cir::CmpOp op,
                                     loweredRhs);
 }
 
+mlir::Value ItaniumCXXABI::lowerMethodCmp(cir::CmpOp op, mlir::Value loweredLhs,
+                                          mlir::Value loweredRhs,
+                                          mlir::OpBuilder &builder) const {
+  assert(op.getKind() == cir::CmpOpKind::eq ||
+         op.getKind() == cir::CmpOpKind::ne);
+
+  cir::IntType ptrdiffCIRTy = getPtrDiffCIRTy(LM);
+  mlir::Value ptrdiffZero = builder.create<cir::ConstantOp>(
+      op.getLoc(), ptrdiffCIRTy, cir::IntAttr::get(ptrdiffCIRTy, 0));
+
+  mlir::Value lhsPtrField = builder.create<cir::ExtractMemberOp>(
+      op.getLoc(), ptrdiffCIRTy, loweredLhs, 0);
+  mlir::Value rhsPtrField = builder.create<cir::ExtractMemberOp>(
+      op.getLoc(), ptrdiffCIRTy, loweredRhs, 0);
+  mlir::Value ptrCmp = builder.create<cir::CmpOp>(op.getLoc(), op.getKind(),
+                                                  lhsPtrField, rhsPtrField);
+  mlir::Value ptrCmpToNull = builder.create<cir::CmpOp>(
+      op.getLoc(), op.getKind(), lhsPtrField, ptrdiffZero);
+
+  mlir::Value lhsAdjField = builder.create<cir::ExtractMemberOp>(
+      op.getLoc(), ptrdiffCIRTy, loweredLhs, 1);
+  mlir::Value rhsAdjField = builder.create<cir::ExtractMemberOp>(
+      op.getLoc(), ptrdiffCIRTy, loweredRhs, 1);
+  mlir::Value adjCmp = builder.create<cir::CmpOp>(op.getLoc(), op.getKind(),
+                                                  lhsAdjField, rhsAdjField);
+
+  // We use cir.select to represent "||" and "&&" operations below:
+  //   - cir.select if %a then %b else false => %a && %b
+  //   - cir.select if %a then true else %b  => %a || %b
+  // TODO: Do we need to invent dedicated "cir.logical_or" and "cir.logical_and"
+  // operations for this?
+  auto boolTy = cir::BoolType::get(op.getContext());
+  mlir::Value trueValue = builder.create<cir::ConstantOp>(
+      op.getLoc(), boolTy, cir::BoolAttr::get(op.getContext(), boolTy, true));
+  mlir::Value falseValue = builder.create<cir::ConstantOp>(
+      op.getLoc(), boolTy, cir::BoolAttr::get(op.getContext(), boolTy, false));
+  auto create_and = [&](mlir::Value lhs, mlir::Value rhs) {
```

Comment:
given both of these lambdas are only used once, is it worth adding them at all?

Reply:
I agree it's unnecessary to add these new ops. Resolving this.